pkg/reconciler/broker/broker.go

```diff
@@ -0,0 +1,499 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package broker
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	brokerv1beta1 "github.com/google/knative-gcp/pkg/apis/broker/v1beta1"
+	"github.com/google/knative-gcp/pkg/broker/config"
+	"github.com/google/knative-gcp/pkg/broker/config/memory"
+	brokerreconciler "github.com/google/knative-gcp/pkg/client/injection/reconciler/broker/v1beta1/broker"
+	brokerlisters "github.com/google/knative-gcp/pkg/client/listers/broker/v1beta1"
+	gpubsub "github.com/google/knative-gcp/pkg/gclient/pubsub"
+	"github.com/google/knative-gcp/pkg/reconciler"
+	"github.com/google/knative-gcp/pkg/reconciler/broker/resources"
+	"github.com/google/knative-gcp/pkg/utils"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/eventing/pkg/apis/eventing"
+	"knative.dev/eventing/pkg/logging"
+	"knative.dev/eventing/pkg/reconciler/names"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/controller"
+	pkgreconciler "knative.dev/pkg/reconciler"
+	"knative.dev/pkg/system"
+)
+
+const (
+	// Name of the corev1.Events emitted from the Broker reconciliation process.
+	brokerReconcileError = "BrokerReconcileError"
+	brokerReconciled     = "BrokerReconciled"
+	brokerFinalized      = "BrokerFinalized"
+	topicCreated         = "TopicCreated"
+	subCreated           = "SubscriptionCreated"
+	topicDeleted         = "TopicDeleted"
+	subDeleted           = "SubscriptionDeleted"
+
+	targetsCMName         = "broker-targets"
+	targetsCMKey          = "targets"
+	targetsCMResyncPeriod = 10 * time.Second
+
+	ingressServiceName = "broker-ingress"
+)
+
+// TODO
+// idea: assign broker resources to cell (configmap) in webhook based on a
+// global configmap (in controller's namespace) of cell assignment rules, and
+// label the broker with the assignment. Controller uses these labels to
+// determine which configmap to create/update when a broker is reconciled, and
+// to determine which brokers to reconcile when a configmap is updated.
+// Initially, the assignment can be static.
+
+// TODO bug: if broker is deleted first, triggers can't be reconciled
+// TODO: verify that if topic is deleted from gcp it's recreated here
+
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	triggerLister   brokerlisters.TriggerLister
+	configMapLister corev1listers.ConfigMapLister
+	endpointsLister corev1listers.EndpointsLister
+
+	// CreateClientFn is the function used to create the Pub/Sub client that interacts with Pub/Sub.
+	// This is needed so that we can inject a mock client for UTs purposes.
+	CreateClientFn gpubsub.CreateFn
```

Comment:
It seems you can directly use the Client here as it's an interface.

Reply:
This is copied from other similar controllers in this repo. I think it makes sense to be consistent with existing controllers here, and open a separate issue if we want to change it everywhere.

Reply:
Sure. We can do some cleanups later on. I